sktime/series_as_features/base/estimators/shapelet_based/_rocket_estimator.py

```diff
@@ -0,0 +1,163 @@
+# -*- coding: utf-8 -*-
+""" RandOm Convolutional KErnel Transform (ROCKET)
+"""
+
+__author__ = "Raven Rudi"
+# TODO
+__all__ = ["BaseROCKETEstimator"]
+
+import numpy as np
+from sklearn.pipeline import make_pipeline
+from sklearn.utils import check_random_state
+from sklearn.utils.multiclass import class_distribution
+
+from sktime.transformations.panel.rocket import Rocket
+from sktime.utils.validation.panel import check_X
+from sktime.utils.validation.panel import check_X_y
+
+
+class BaseROCKETEstimator:
+    """
+    Base class for ROCKET classifier and ROCKET regressor.
+
+    Allows the creation of an ensemble of ROCKET estimators to allow for
+    generation of probabilities as the expense of scalability.
+
+    Parameters
+    ----------
+    num_kernels             : int, number of kernels for ROCKET transform
+    (default=10,000)
+    ensemble                : boolean, create ensemble of ROCKET's (default=False)
+    ensemble_size           : int, size of the ensemble (default=25)
+    random_state            : int or None, seed for random, integer,
+    optional (default to no seed)
+
+    Attributes
+    ----------
+    classifiers             : array of IndividualTDE classifiers
+    weights                 : weight of each classifier in the ensemble
+    weight_sum              : sum of all weights
+    n_classes               : extracted from the data
+
+    Notes
+    -----
+    @article{dempster_etal_2019,
+      author  = {Dempster, Angus and Petitjean, Francois and Webb,
+      Geoffrey I},
+      title   = {ROCKET: Exceptionally fast and accurate time series
+      classification using random convolutional kernels},
+      year    = {2019},
+      journal = {arXiv:1910.13051}
+    }
+
+    """
+
+    # Capability tags
+    capabilities = {
+        "multivariate": True,
+        "unequal_length": False,
+        "missing_values": False,
+    }
+
+    def __init__(
+        self,
+        num_kernels=10000,
+        ensemble=False,
+        ensemble_size=25,
+        random_state=None,
+    ):
+        self.num_kernels = num_kernels
+        self.ensemble = ensemble
+        self.ensemble_size = ensemble_size
+        self.random_state = random_state
+
+        self.classifiers = []
+        self.weights = []
+        self.weight_sum = 0
+
+        self.n_classes = 0
+        self.classes_ = []
+        self.class_dictionary = {}
+
+        # We need to add is-fitted state when inheriting from scikit-learn
+        self._is_fitted = False
+
+    def fit(self, base_model, X, y):
```

Comment:
rather than passing the `base_model` here as an argument I would define it as a class variable (see TSF refactoring for an example)

Reply:
`fit` is part of the common API, so we shouldn't change its signature

Reply:
I solved that by making `BaseROCKETEstimator` an abstract class with a `base_estimator` as an abstract property. So the child classes are forced to provide a `base_estimator`. In the tsf refactoring, the parent class implicitly assumes that the child class provides a `base_estimator` . I don't think this is a neat structure :)